ksql-serde/src/main/java/io/confluent/ksql/serde/json/KsqlJsonDeserializer.java

```diff
@@ -110,7 +119,58 @@ private Struct toStruct(final byte[] rowJsonBytes) {
     return struct;
   }
 
-  private Struct fromMap(final Map<String, Object> valueMap) {
+  /**
+   * Should this value be treated as a record or a top level map?
+   *
+   * <p>The deserializer supports deserializing top-level unnamed primitives, arrays and maps if
+   * the target schema only contains a single field of the required type, e.g. the value '10' can be
+   * deserialized into a schema containing a single numeric field. Likewise, the value '[1,2]' can
+   * be deserialized into a schema containing a single array field. The same is true of maps, but
+   * maps can be ambiguous.
+   *
+   * <p>There is ambiguity when handling top level maps due to the fact that JSON does not have a
+   * map type, so KSQL persists maps as JSON objects.  JSON objects are also used to represent KSQL
+   * structs. When the deserializer is presented with a JSON object and a target schema that
+   * contains a single MAP field, it is not immediately apparent if this should be deserialized as a
+   * struct or a top-level map.
+   *
+   * <p>Map fields are only ambiguous if:
+   * <ul>
+   *   <li>The target schema contains only a single field</li>
+   *   <li>and that field is a map</li>
+   *   <li>and that map has a string key</li>
+   * </ul>
+   *
+   * <p>This method determines which it should be treated as by comparing the target schema with
+   * the actual value.
+   *
+   * <p>Note: It is strongly recommended that users avoid the situation by ensuring that single
+   * map fields never have the same name as any of the keys in the map.
+   *
+   * @param map the map to check.
+   * @return true if it should be treated as a record.
+   */
+  private boolean treatAsRecord(final Map<String, Object> map) {
+    if (!ambiguousMapSchema) {
+      return true;
+    }
+
+    final Field onlyField = schema.fields().get(0);
+    final String fieldName = onlyField.name();
+    final Object entry = map.entrySet().stream()
+        .filter(e -> fieldName.equalsIgnoreCase(e.getKey()))
+        .findFirst()
```

Comment:
I think we want to find the first that is coercible here, e.g.:

```
final Object entry = map.entrySet().stream()
    .filter(e -> fieldName.equalsIgnoreCase(e.getKey())
    .filter(e -> SerdeUtils.isCoercible(entry, onlyField.schema())
    .findFirst()
    .map(Entry::getValue)
    .orElse(null);
```

That way, if you have a schema like `FOO MAP<STRING, STRING>` and a value like `{"foo": "bar", "fOO": {"a": "b"}}`, we deserialize as a record (which I think is what we want?)

Reply:
Good call.